src/main.cpp

```diff
@@ -95,106 +89,59 @@ int main(int argc, char *argv[])
   if (use_float)
     run<float>();
   else
-    run<double>(); 
+    run<double>();
 
+  return EXIT_SUCCESS;
 }
 
+// Returns duration of executing function f:
+template <typename F>
+double time(F&& f) {
+  using clk_t = std::chrono::high_resolution_clock;
+  using dur_t = std::chrono::duration<double>;
+  auto start = clk_t::now();
+  f();
+  return dur_t(clk_t::now() - start).count();
+}
 
-// Run the 5 main kernels
+// Run specified kernels
 template <typename T>
-std::vector<std::vector<double>> run_all(Stream<T> *stream, T& sum)
+std::vector<std::vector<double>> run_all(std::unique_ptr<Stream<T>>& stream, T& sum)
 {
+  // Times for each measured benchmark:
+  std::vector<std::vector<double>> timings(num_benchmarks);
 
-  // List of times
-  std::vector<std::vector<double>> timings(5);
-
-  // Declare timers
-  std::chrono::high_resolution_clock::time_point t1, t2;
+  // Time a particular benchmark:
+  auto dt = [&](Benchmark const& b)
+  {
+    switch(b.id) {
+    case BenchId::Copy:    return time([&] { stream->copy(); });
+    case BenchId::Mul:     return time([&] { stream->mul(); });
+    case BenchId::Add:     return time([&] { stream->add(); });
+    case BenchId::Triad:   return time([&] { stream->triad(); });
+    case BenchId::Dot:     return time([&] { sum = stream->dot(); });
+    case BenchId::Nstream: return time([&] { stream->nstream(); });
+    default:
+      std::cerr << "Unimplemented benchmark: " << b.label << std::endl;
+      abort();
+    }
+  };
 
   // Main loop
-  for (unsigned int k = 0; k < num_times; k++)
+  for (size_t i = 0; i < num_benchmarks; ++i)
```

Comment:
I think this is the main outstanding change on this PR. We need the classic kernels (`BenchId::Classic`) to run as:
```
for (size_t k = 0; k < num_times; k++) {
  for (size_t i = 0; i < num_benchmarks; ++i) {
    dt(bench[i])
  }
}
```
rather than the current order.
Therefore, the correct ordering runs:
`Copy -> Mul -> Add -> Triad -> Dot -> Copy -> Multiple -> ...`